### PR TITLE
Implement `jsctl  cluster backup` command

### PR DESCRIPTION
**Pull request by charlieegan3**
This implements a command to generate a cluster back up from the state in the K8S API.

Example output can be found in [internal/kubernetes/backup/fixtures/backup.yaml](https://github.com/jetstack/jsctl/blob/629b1a930c5bc9a4a36f2c53e5a70ae3c40c07a5/internal/kubernetes/backup/fixtures/backup.yaml), status and createdTime values come from the k8s yaml generation. These can be removed with a further round of redaction if we'd like.

### CODE REVIEW

**Review by irbekrm — CHANGES_REQUESTED** [review]
Thanks Charlie, some great work on this again 👏🏼 

I've left a couple comments most notably
- I think this command too should go in experimental section (as again it will most likely be used in production cases)
- I think we should back up cert-manager alpha/beta resources
- (unlike what we discussed before) I think we should always just not back up ingress certs, at least I cannot see why it would ever work backing them up

I've not tried the backup yet, but will do when the alpha/beta backup is added if you agree with doing that

File: internal/kubernetes/backup/backup.go [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+				}
+			}
+		}
+		if !skip {
+			clusterBackup = append(clusterBackup, c)
+		}
+	}
+
+	// fetch certificate request policies
+	certificateRequestPolicyClient, err := clients.NewCertificateRequestPolicyClient(opts.RestConfig)
```
[the remark this answers is not in the document]
Comment (charlieegan3) [reply internal/kubernetes/backup/backup.go:114]: Addressed in 83283775b2c0b283d4003d06b416dc9a9f4d6a52

File: internal/kubernetes/backup/backup.go [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+	if err != nil {
+		return &ClusterBackup{}, fmt.Errorf("failed to list certificates: %w", err)
+	}
+	for _, c := range certificates.Items {
+		// if we're not including ingress certificates, skip them
+		skip := false
+		if !opts.IncludeIngressCertificates && len(c.OwnerReferences) > 0 {
+			for _, owner := range c.OwnerReferences {
+				if owner.Kind == "Ingress" {
+					skip = true
```
[the remark this answers is not in the document]
Comment (charlieegan3) [reply internal/kubernetes/backup/backup.go:101]: Addressed in 330cdfaa10d2d6098c39d8408c5857c6a5db984b

File: internal/command/clusters/backup.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+func Backup(run types.RunFunc, kubeConfigPath string) *cobra.Command {
+	var formatResources bool
+	var outputFormat string
+
+	var includeCertificates bool
+	var includeIssuers bool
+	var includeCertificateRequestPolicies bool
+
+	var skipIngressCerts bool
```
[the remark this answers is not in the document]
Comment (charlieegan3) [reply internal/command/clusters/backup.go]: Addressed in 21c8d8aa307ba8aacbbba8e9b08f6e8b7923b14d and 330cdfa

**Comment by charlieegan3** [comment]
> I think this command too should go in experimental section

Addressed in 28f181698b03f2d23afdfa70db4de2dfd333bae8

**Review by charlieegan3 — COMMENTED** [review]
> I think this command too should go in experimental section

Addressed in 28f181698b03f2d23afdfa70db4de2dfd333bae8

**Comment by charlieegan3** [comment]
I've addressed the other comments, I'm going to have a think about how we manage multiple API versions now.

File: internal/kubernetes/backup/backup.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (charlieegan3) [reply internal/kubernetes/backup/backup.go:114]: I got confused, how are they not to be used? Won't they come back as part of `jsctl cluster restore`?

File: internal/kubernetes/backup/backup.go [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+	}
+	clusterBackup = append(clusterBackup, issuers...)
+
+	// fetch certifcates
+	certificateClient, err := clients.NewCertificateClient(opts.RestConfig)
+	if err != nil {
+		return &ClusterBackup{}, fmt.Errorf("failed to create client for certificates: %w", err)
+	}
+
+	var certificates v1certmanager.CertificateList
```
[the remark this answers is not in the document]
Comment (charlieegan3) [reply internal/kubernetes/backup/backup.go:109]: As discussed, addressed in afacf30b4106c43df819ebce522717c75aa94cf8 by ejecting.

File: internal/command/clusters/backup.go [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+		}),
+	}
+
+	flags := cmd.PersistentFlags()
+	flags.BoolVar(&formatResources, "format-resources", true, "if set, will remove some fields from resources such as status and metadata to allow them to be cleanly applied later")
+	flags.StringVar(&outputFormat, "format", "yaml", "output format, one of: yaml, json")
+
+	flags.BoolVar(&includeCertificates, "include-certificates", true, "if set, certificate resources will be included in the backup")
+	flags.BoolVar(&includeIssuers, "include-issuers", true, "if set, issuer resources will be included in the backup")
+	flags.BoolVar(&includeCertificateRequestPolicies, "include-certificate-request-policies", true, "if set, certificate request policy resources will be included in the backup")
```
[the remark this answers is not in the document]
Comment (charlieegan3) [reply internal/command/clusters/backup.go:65]: As discussed, we have opted to eject in this case instead. afacf30b4106c43df819ebce522717c75aa94cf8 does this.

File: internal/command/clusters/backup.go [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+			return nil
+		}),
+	}
+
+	flags := cmd.PersistentFlags()
+	flags.BoolVar(&formatResources, "format-resources", true, "if set, will remove some fields from resources such as status and metadata to allow them to be cleanly applied later")
+	flags.StringVar(&outputFormat, "format", "yaml", "output format, one of: yaml, json")
+
+	flags.BoolVar(&includeCertificates, "include-certificates", true, "if set, certificate resources will be included in the backup")
```
[the remark this answers is not in the document]
Comment (charlieegan3) [reply internal/command/clusters/backup.go]: See afacf30b4106c43df819ebce522717c75aa94cf8. I hope this addresses as we discussed.

File: internal/kubernetes/backup/backup.go [hunk 7]
```diff
@@ -69,6 +70,29 @@ func FetchClusterBackup(ctx context.Context, opts ClusterBackupOptions) (*Cluste
...
+	}
+	for _, crd := range crds.Items {
+		if crd.Spec.Group != "cert-manager.io" {
+			continue
+		}
+		if len(crd.Spec.Versions) == 0 {
+			return nil, fmt.Errorf("unexpectedly found no versions on cert-manager.io CRD %s", crd.Name)
+		}
+		// the first version in the list must be v1 as we expect v1 resources
+		if crd.Spec.Versions[0].Name != "v1" {
```
Comment (irbekrm) [inline internal/kubernetes/backup/backup.go:91]: Sorry @charlieegan3 I wasn't very clear on this. 
An unsupported cert-manager.io CRD would be one that does not have `v1` version. It is fine if they have `v1` *as well as* some of the alpha and beta versions- the CRDs between cert-manager v1 - v1.6 had all these versions, see i.e [this CRD](https://github.com/cert-manager/cert-manager/releases/download/v1.4.1/cert-manager.crds.yaml). It's not a problem for the backup command because as long as there is a v1 version, there will be version conversion happening and if we get custom resources at v1, anything that user created that is i.e `v1alpha2` will also be returned, so the backup will not miss any resources.

Let me know if this makes sense. Basically this check should just ensure that `v1` is one of the versions there-it's ok if other versions are also present.

**Review by irbekrm — CHANGES_REQUESTED** [review]
Thanks for making all the changes!

I've not ran all the commands yet, but took a look at the CRD version check and I think it's close 👍🏼

File: internal/command/clusters/backup.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (charlieegan3) [reply internal/command/clusters/backup.go:75]: I was wondering about this too. I thought that it needed to be v1, and have v1 with served: true (though I don't check that here).

My perhaps naive guess was that if v1 was not the first version, then the first version would be a later version, e.g. v2, and this would be served instead. Given this felt like a long way off and would create issuers for us, I thought we could just check the version of the first version item.

In summary, I think that if we have a v1 in the list, but it's not served, then we might have an issue?

File: internal/kubernetes/backup/backup.go (hunk 7 above)
Comment (charlieegan3) [reply internal/kubernetes/backup/backup.go:91]: I was also wondering about this. I have done a little looking and I think you're right, but that v1 would also need to have `served: true` set. I'll update this now.

Comment (charlieegan3) [reply internal/kubernetes/backup/backup.go]: I think that 7419e678b47e0a2dbb137700c3f5b23294d3fff1 addresses this now 🤞

Comment (irbekrm) [reply internal/kubernetes/backup/backup.go]: Thanks, I've tested with cert-manager 0.16 that only has alpha and beta versions and can seee that it works.

**Review by irbekrm — CHANGES_REQUESTED** [review]
I've ran the command now and it seems to largely work as expected, a couple more things that might need fixing before we can merge:

- none of the flags seem to work
- if redirected to file, it gets prefixed with issuer GVK string so is not a valid yaml that could be applied

**Comment by charlieegan3** [comment]
> none of the flags seem to work

Oops, in e0bd75308dba86c77f8af81fb9f9ab4b4f605034 I use the options in the back up function. I have also rebased this to fix the issue with stdout etc not being passed down correctly.

> if redirected to file, it gets prefixed with issuer GVK string so is not a valid yaml that could be applied

e0bd75308dba86c77f8af81fb9f9ab4b4f605034 also addresses this.

**Review by irbekrm — CHANGES_REQUESTED** [review]
Thanks again for working on this Charlie, I think the command looks great now.

I've ran a couple examples and most things look good.

I noticed that `--format=json` flag is not working. Do you think it'd be easy to get that fixed? It would be nice to have this flag as that way we can guide users to use `jq` (which I find to be a better tool than `yq`) when re-applying the necessary resources from backup.
Otherwise we could remove the flag, happy to merge this either way.

**Comment by charlieegan3** [comment]
Hmm, that's a good spot 😬 

I've updated the command so it works like this now:

```
 $ go run main.go experimental clusters backup
---
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  creationTimestamp: null
  name: example-com
  namespace: jetstack-secure
spec:
  dnsNames:
  - example.com
  issuerRef:
    group: cert-manager.io
    kind: Issuer
    name: ca-issuer
  secretName: example-com-tls
status: {}
...arlieegan/Code/jsctl $ go run main.go experimental clusters backup --format=json
{
  "apiVersion": "v1",
  "items": [
    {
      "kind": "Certificate",
      "apiVersion": "cert-manager.io/v1",
      "metadata": {
        "name": "example-com",
        "namespace": "jetstack-secure",
        "creationTimestamp": null
      },
      "spec": {
        "dnsNames": [
          "example.com"
        ],
        "secretName": "example-com-tls",
        "issuerRef": {
          "name": "ca-issuer",
          "kind": "Issuer",
          "group": "cert-manager.io"
        }
      },
      "status": {}
    }
  ],
  "kind": "List"
}...arlieegan/Code/jsctl $ go run main.go experimental clusters backup --format=things
unknown output format: things
exit status 1

```

Lmk if you have any thoughts on 4bb34682b290f8ea2df795e3f48de80269298786 which adds this change. I thought it made sense to wrap it in a k8s list so that it can be kubectl applied.

**Comment by charlieegan3** [comment]
> Why did we need to make this into a list?

I think that is required to allow it to be piped to `kubectl apply -f -`. YAML has `---` but there's no equivalent in JSON files for kubectl afaik - other than wrapping in a v1 list. Note, this is an unversioned list, not an issuerlist.

> At the moment the --experimental-issuers-backup-file flag does not parse a list type

hmm, I can update that to support this unversioned JSON list.

**Comment by irbekrm** [comment]
> hmm, I can update that to support this unversioned JSON list.

yeah I've not thought about this in depth, but the output file should be usable to by the command that converts issuers.

**Review by irbekrm — APPROVED** [review]
Thanks a lot for working on this Charlie

I've tested the --experimental-issuers-backup-file flag with JSON output from this command and saw that all works as expected

/lgtm

**Comment by charlieegan3** [comment]
Thanks Irbe!
